### PR TITLE
Enable updating Gravatars on Desktop

### PR DESCRIPTION
**Pull request by v18**
This PR enables the feature flag "me/edit-gravatar" on Desktop.

@designsimply, this is now ready for testing 🎉 thank you :) :)
@astralbodies - could you please let me know if there's anything else I need to do to get this in to the June desktop release? A: no, this is good to go.

# Testing Instructions

## Install the desktop app
1. Install the desktop app using [these instructions](https://github.com/Automattic/wp-desktop#getting-started--running-locally). The full instructions are also listed below:
`git clone git@github.com:Automattic/wp-desktop.git`
`git submodule init`
Create a `calypso/config/secrets.json` file and fill it with secrets. You can find the location of the secrets at PCYsg-6rI-p2, under "Build App"
2. While in `/calypso`, checkout this branch `update/edit-gravatar-on-desktop`
3. In `/wp-desktop` use `make run` to build and run the app

## Seeing the feature
You should be able to see the feature if you go to 'My Profile' by clicking your Gravatar in the top right corner:

screen shot 2017-05-27 at 10 30 27 am

## Things that should work

General
- responsive: resizing the screen should still allow you see and use the feature
- i18n: you should be able to see the feature translated (the 'Change My Photo' string is not yet translated)

Selecting an image
- clicking: click on the image to select a photo from your computer
- drag and drop: you should be able to drag an image onto the circle
- file type: you should see an error if you select an svg file or if your file has an image extension but isn't actually an image

Editing the image
- editing the image: you should be able to use the image editor that comes up after selecting the image to resize it. the dimensions should always be square - you should not be allowed to make a non-square selection
- you should be able to cancel the edit and the image won't get uploaded
- you should be able to click on 'Change My Photo' to save the new image

Uploading
- after clicking 'Change My Photo' you should see a spinner to indicate upload
- after the the upload succeeds, you should see a success notice
- if the upload fails, you should see a failure notice

Unverified users
- if you log in using an unverified user, you should not be able to upload an image, and instead you'll see a message letting you send another verification email
- note: there's a bug (#14538) with the 'Resend Email' button for now, which I'll work on fixing next week

Here's a video showing a bad image, and then a successful upload:

http://cld.wthms.co/eRmtOuw

## Strings

I wanted to make sure all the different strings make sense. Here's a list of all the possible ones and when they can happen:

- "**An image of filetype %s is not allowed**" and "**An image of that filetype is not allowed**" - when a user picks an image type that's not allowed (e.g. svn)
- "**We couldn't save that image, please try another one**" - when the image could not be shown edited in the image editor
- "**We can't save a new Gravatar now. Please try again later.**" - when there's an authentication problem. I expect this to never be shown
- "**Click to change photo**" - the text shown on the button
- "**Verify email first**" -  the text shown for users with an unverifeid email
- "**Drop to upload profile photo**" - the text shown when dragging an image 
- "**Your profile photo is public.**" - the text shown right below the button
- "**The avatar you use on WordPress.com comes from <a>Gravatar</a>  - a universal avatar service. Your photo may be displayed on other sites where you use your email address your@email.com**" - the explanation text shown in a tooltip if the user clicks the information button
- "**New Gravatar was not saved.**" - the error message shown if the Gravatar was not saved on Gravatar.com
- "**New Gravatar uploaded successfully!**" - success message shown when Gravatar is saved on Gravatar.com

Adding the 'Needs Copy Review' label for this.

## Instructions to see the new strings:
1. `export ENABLE_FEATURES=me/edit-gravatar && make run`
2. Log in to WordPress.com as an unverified user, go to `http://calypso.localhost:3000/me`
3. See the "Verify your email" string
4. Click on the Gravatar, see the new email verification strings:
>Please verify your email address:
>[email]
> When you first signed up for a WordPress.com account we sent you an email. Please open it and click on the blue button to verify your email address.
>Verifying your email allows us to assist you if you ever lose access to your account in the future.
5. Click the information button (i), see the new strings:
> The avatar you use on WordPress.com comes from Gravatar, a universal avatar service (it stands or "Global Avatar," get it?).
> Your image may also appear on other sites using Gravatar whenever you're logged in with your email address [email]
6. Log out, and log back in as a verified user. Go to `http://calypso.localhost:3000/me`
7. Click on the Gravatar and choose a new photo. See the error message "Hmm, we can't save a new Gravatar now. Please try again later, or contact our Happiness team."
8. Set your bearer token in the console using `localStorage.setItem('bearerToken', 'your_bearer_token');` You can [find your bearer token using these instructions](https://github.com/Automattic/wp-calypso/pull/13159#issuecomment-303583833), or ask me for a test username + bearer token.
9. Click on the Gravatar and try to upload a png photo, and see the string 'Change My Photo'
10. Click on 'Change My Photo', see the image upload, and see the string "You successfully uploaded a new Gravatar — looking sharp!" in the success message
11. Try to upload an svg photo and see the string "Sorry, svg files are not supported — please make sure your image is in JPG, GIF, or PNG format."
12. Try to upload this non-photo file, and see the error message "Sorry, images of that filetype are not supported — please make sure your image is in JPG, GIF, or PNG format."
13. Turn off networking before clicking 'Change My Photo', and see this error message "Hmm, your new Gravatar was not saved. Please try uploading again, or contact our Happiness team."
14. On line 116 of `/client/blocks/edit-gravatar/index.jsx` add `error = true`, refresh the page and try to upload an image. See the error message "We couldn't save that image — please try another one, or contact our Happiness team."

### CODE REVIEW

**Comment by michelleweber** [comment]
A few tweaks for the copy for tone (to line these up with errors/messaging we've been updating in other parts of Calypso/on mobile) and to try and guide users a bit more, and a few questions.

- "**Sorry, %s files are not supported -- please make sure your image is in JPG, GIF, or PNG format.**" and "**Sorry, images of that filetype are not supported -- please make sure your image is in JPG, GIF, or PNG format.**" 
- "**We couldn't save that image -- please try another one.**" -- When does this happen? It would be good to be able to provide a detail, so the user knows why the image couldn't save and can make a better selection.
- "**Hmm, we can't save a new Gravatar now. Please try again later.**" 
- "**Click to change photo**" 
- "**Wait! Please verify your email address first -- check your email for a verification message from WordPress.com.**" -  I'm assuming this one means they haven't yet verified with wp.com, so I edited it accordingly.
- "**Drop to upload profile photo**" 
- "**Your profile photo is public.**" 
- "**The avatar you use on WordPress.com comes from <a>Gravatar</a>, a universal avatar service (it stands or "Global Avatar," get it?). Your image may also appear on other sites that user Gravatar if you register with the email address your@email.com**" 
- "**Hmm, your new Gravatar was not saved. Please try uploading again.**" - Is this our fault? Their fault? What should they do next?
- "**You successfully uploaded a new Gravatar -- looking sharp!**"

**Comment by v18** [comment]
@michelleweber thank you for your edits! I have a question and some answers:

For "Verify email first" -> "Wait! Please verify your email address first...", it needs to be a short string to fit in the button we have. It looks like this right now:



Is that string ok, or is there a better short version?

When they click on the image, they see this message:



To answer your questions:

> When does this happen? It would be good to be able to provide a detail, so the user knows why the image couldn't save and can make a better selection.

This would happen due to an unknown error when the image can't be used in the image editor. If it happens, they should try again, or contact support who'd help them troubleshoot and hopefully alert us to the error so we can either fix it or edit the error message.

> Is this our fault? Their fault? What should they do next?

This would be our fault when trying to upload a Gravatar to Gravatar.com. It could happen for a number of reasons: a bad filetype was able to be uploaded anyway, a server error (like 400 or 500), etc. Right now, I don't show the error from the server, but I track if we display an error.

The user would either need to try again, or contact support. If we find that a certain error is happening a lot, we could either fix it, or show a better error message.

**Comment by v18** [comment]
@michelleweber - I have a couple more questions :)

1. There's one more string I forgot last time: "Change My Photo". It appears on the button right before saving the new image:

screen shot 2017-06-01 at 8 27 27 am

2. The text explaining Gravatar says "if you register with", but it also appears for users who have already registered like this:

screen shot 2017-06-01 at 8 29 13 am

Should it be updated to be more general & include users who have already registered?

**Comment by benhuberman** [comment]
Hey @v18 -- here are some thought on your most recent comments/questions:
- I'd consider switching "Verify email first" to "Verify your email" -- I think it could increase the chance of people clicking on this to see the additional context/information (whereas "Verify email first" might leave them hanging, not sure what to do next).
- The email verification message contains some missing spaces, and I think could be made just a tad clearer, e.g.:
> Please verify your email address:
[email address]
When you first signed up for a WordPress.com account we sent you an email. Please open it and click on the blue button to verify your email address.
Verifying your email allows us to assist you if you ever lose access to your account in the future.

Re: image upload failures:
> they should try again, or contact support who'd help them troubleshoot and hopefully alert us to the error so we can either fix it or edit the error message

I'd consider tweaking the copy to reflect these options: 
**We couldn't save that image -- please try another one, or contact our Happiness team.**
(with a link to /help maybe? Basically, I'd rather we don't create dead ends whenever possible.)

Same goes for this:

> Hmm, your new Gravatar was not saved. Please try uploading again, **or contact our Happiness team.**

Re: the additional string:

> There's one more string I forgot last time: "Change My Photo". It appears on the button right before saving the new image:

This looks/sounds good to me!

Re: Gravatar on other sites:
> Should it be updated to be more general & include users who have already registered?

I'd say yes. Maybe something like...

**Your image may also appear on other sites using Gravatar whenever you're logged in with your email address [email address].**

Would this make sense?

**Comment by v18** [comment]
Hi @benhuberman! Thank you so much for your feedback! 🎉  I updated all the strings to what you suggested.

> with a link to /help maybe?

Doesn't look like we're able to link directly from the notifications yet (so maybe work for another PR). In addition, I'm not 100% sure if we're limiting links to support in any way (paid users etc), so this requires more digging before implementing.

File: client/blocks/edit-gravatar/index.jsx [hunk 1]
```diff
@@ -204,7 +208,7 @@ export class EditGravatar extends Component {
 		const icon = ( user.email_verified ? 'cloud-upload' : 'notice' );
 		const buttonText = ( user.email_verified
 			? translate( 'Click to change photo' )
-			: translate( 'Verify email first' )
+			: translate( 'Verify your email' )
```
Comment (a8ci18n) [inline client/blocks/edit-gravatar/index.jsx]: Hi! I've found a possible matching string that has already been translated **24** times:
`translate( 'Verify your email address' )` *ES Score: 9*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/edit-gravatar-on-desktop)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/components/email-verification/email-verification-dialog/index.jsx [hunk 2]
```diff
@@ -56,13 +56,17 @@ class VerifyEmailDialog extends Component {
 
 	render() {
 		const strings = {
-			confirmHeading: this.props.translate( 'Please confirm your email address' ),
+			confirmHeading: this.props.translate( 'Please verify your email address:' ),
```
Comment (a8ci18n) [inline client/components/email-verification/email-verification-dialog/index.jsx]: Hi! I've found a possible matching string that has already been translated **28** times:
`translate( 'Please Verify Your Email Address' )` *ES Score: 13*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/edit-gravatar-on-desktop)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by benhuberman** [comment]
> Doesn't look like we're able to link directly from the notifications yet (so maybe work for another PR). In addition, I'm not 100% sure if we're limiting links to support in any way

Gotcha, @v18. I'll defer to you in terms of what makes the most sense and/or is currently feasible. I (and I think Michelle too, in her initial comment) mostly wanted to flag this instance where users might be facing an unclear path forward, which could be frustrating. Also: if they can't actually contact support (when they're not on a plan that offers that) then it's likely not a good idea to suggest they do.

**Comment by v18** [comment]
Thanks Ben! I removed the happiness mention for now.

File: client/blocks/edit-gravatar/index.jsx [hunk 3]
```diff
@@ -136,7 +138,7 @@ export class EditGravatar extends Component {
 			uploadGravatarAction( imageBlob, bearerToken, user.email );
 		} else {
 			receiveGravatarImageFailedAction( {
-				errorMessage: translate( "We can't save a new Gravatar now. Please try again later." ),
+				errorMessage: translate( "Hmm, we can't save a new Gravatar now. Please try again later." ),
```
Comment (a8ci18n) [inline client/blocks/edit-gravatar/index.jsx]: :ok: This change is safe, we'll keep the existing 17 translations.

File: client/blocks/edit-gravatar/index.jsx [hunk 4]
```diff
@@ -116,7 +118,7 @@ export class EditGravatar extends Component {
 
 		if ( error ) {
 			receiveGravatarImageFailedAction( {
-				errorMessage: translate( "We couldn't save that image, please try another one" ),
+				errorMessage: translate( "We couldn't save that image — please try another one." ),
```
Comment (a8ci18n) [inline client/blocks/edit-gravatar/index.jsx]: :ok: This change is safe, we'll keep the existing 17 translations.

**Pull request merged**